### PR TITLE
⭐ Report integration status

### PR DESCRIPTION
**Pull request by imilchev**
- [x] Report integration status
- [x] Add unit tests

### CODE REVIEW

**Comment by czunker** [comment]
This fixes #394

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "gopls": {
```
Comment (imilchev) [inline .vscode/settings.json:2]: we already use this setting in the Mondoo repo. It does change the formatting a bit. I am adding it here so we share the settings

File: tests/framework/utils/k8s_helper.go [hunk 2]
```diff
@@ -334,7 +333,7 @@ func (k8sh *K8sHelper) createTestLogFile(name, namespace, testName, suffix strin
 	dir, _ := os.Getwd()
 	logDir := path.Join(dir, "_output/tests/")
 	if _, err := os.Stat(logDir); os.IsNotExist(err) {
-		err := os.MkdirAll(logDir, 0777)
+		err := os.MkdirAll(logDir, 0o777)
```
Comment (czunker) [inline tests/framework/utils/k8s_helper.go:336]: Why this change?

I also saw it in some other files for the `DefaultMode` of secrets.

**Comment by czunker** [comment]
Nice! :tada:

**Review by czunker — APPROVED** [review]

File: pkg/mondooclient/client.go [hunk 3]
```diff
@@ -364,3 +381,65 @@ func (s *mondooClient) IntegrationCheckIn(ctx context.Context, in *IntegrationCh
 
 	return out, nil
 }
+
+type ReportStatusRequest struct {
+	Mrn string `protobuf:"bytes,1,opt,name=mrn,proto3" json:"mrn,omitempty"`
+	// this is the status of the integration itself (is it active/checking in, errored, etc)
+	Status Status `protobuf:"varint,2,opt,name=status,proto3,enum=mondoo.integrations.v1.Status" json:"status,omitempty"`
+	// this can be any information about the current state of the integration. it will be displayed to the user as-is where supported
+	LastState interface{} `protobuf:"bytes,4,opt,name=last_state,json=lastState,proto3" json:"last_state,omitempty"`
```
[the remark this answers is not in the document]
Comment (imilchev) [reply pkg/mondooclient/client.go:390]: this is copy-pasted code from the mondoo repo because we cannot generate those files ourselves at the moment. Ignore all the types defined here

File: pkg/mondooclient/client.go [hunk 4]
```diff
@@ -364,3 +381,65 @@ func (s *mondooClient) IntegrationCheckIn(ctx context.Context, in *IntegrationCh
...
+	// Allows the agent to report its current version
+	Version string `protobuf:"bytes,5,opt,name=version,proto3" json:"version,omitempty"`
+	// messages that convey extra information about the integration - these messages can be informational, warnings or errors. Can be used
+	// to report non-critical errors/warnings without neccesarily changing the whole integration status.
+	Messages []IntegrationMessage `protobuf:"bytes,7,opt,name=messages,proto3" json:"messages,omitempty"`
```
[the remark this answers is not in the document]
Comment (imilchev) [reply pkg/mondooclient/client.go:395]: same

File: tests/framework/utils/k8s_helper.go (hunk 2 above)
Comment (imilchev) [reply tests/framework/utils/k8s_helper.go:336]: I believe the `gofumpt` linter does that because the other notation is not accepted

**Comment by imilchev** [comment]
@joelddiaz since you did the console integration initially it might be good to have a look at this PR as well to make sure I haven't missed something

File: token [hunk 5]
```diff
@@ -0,0 +1 @@
+abcdefgh
```
Comment (joelddiaz) [inline token:1]: This some leftover local debug/test file?

**Review by joelddiaz — COMMENTED** [review]
One thing we've discussed previously was sending back either the full or a piece of the MondooAuditConfig so that the UI can do things like say "your last reported configuration differs from the configuration stored in Mondoo".

Assuming that is still desired, we may want to send along enough of the MondooAuditConfig so that we can do a server-side determination when the running config doesn't match the desired config.

Also, (maybe a question for @vjeffrey ): I was under the impression that we could use the ReportStatus() call instead of the CheckIn() call. Therefore I just expected the CheckIn()s to basically go away and be replaced by ReportStatus() calls instead. Not that the approach in the PR is wrong, but just for my knowledge, is there any particular reason do keep the CheckIn() as-is and adding this new codepath with the ReportStatus() calls?

**Comment by imilchev** [comment]
> One thing we've discussed previously was sending back either the full or a piece of the MondooAuditConfig so that the UI can do things like say "your last reported configuration differs from the configuration stored in Mondoo".
> 
> 
> 
> Assuming that is still desired, we may want to send along enough of the MondooAuditConfig so that we can do a server-side determination when the running config doesn't match the desired config.
> 
> 
> 
> Also, (maybe a question for @vjeffrey ): I was under the impression that we could use the ReportStatus() call instead of the CheckIn() call. Therefore I just expected the CheckIn()s to basically go away and be replaced by ReportStatus() calls instead. Not that the approach in the PR is wrong, but just for my knowledge, is there any particular reason do keep the CheckIn() as-is and adding this new codepath with the ReportStatus() calls?

I can explain the reasoning behind this. You are right that ReportStatus implicitly does a CheckIn as well. However, here we still use both because I didn't want to just replace the CheckIn with ReportStatus and report every 10mins. The current implementation  checks in every 10mins and does ReportStatus every time the status actually changes. If there is no change in the status (e.g. config hasn't changed and no errors have occurred), we do not report the status object. For that reason the checkin logic is still there, such that we still do a ping even if there are no status changes

File: cmd/mondoo-operator/operator/cmd.go [hunk 6]
```diff
@@ -74,11 +75,17 @@ func init() {
 			return err
 		}
 
+		v, err := k8s.GetServerVersion(mgr.GetConfig())
```
[the remark this answers is not in the document]
Comment (imilchev) [reply cmd/mondoo-operator/operator/cmd.go:78]: that is also my understanding so I implemented it like this. Basically is you do a full cluster upgrade the operator will be re-scheduled at least once for sure, which means we will get the new version. If you upgrade half of your nodes, then that won't happen but then you are also doing something weird, so I wouldn't try to fix that situation

**Comment by imilchev** [comment]
Processed the comments and rebased on `main`

**Review by joelddiaz — APPROVED** [review]
